west: runners: Add new -i/--dev-id device identifier common runner option

carlescufi opened the pull request:
In order to avoid different runners implementing device identification each in their own way, introduce a new `-i/--dev-id` command-line switch that allows the user to specify which board/debugger/node is to be targeted by the flash or debug operation.

carlescufi committed:
runners: Add new common -i/--dev-id command-line option

In an effort to standardize the way that a particular debugger or device
instance is identified when there are multiple present, introduce a new
-i/--dev-id option common to all runners that allows the user to specify
which device to interact with when there are multiple connected.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
runners: jlink: Convert to -i/--dev-id

Remove the previous jlink-specific --id option and switch to the common
-i/--dev-id one.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
runners: canopen: Convert to -i/--dev-id

Remove the previous canopen-specific --node-id option and switch to the
common -i/--dev-id one.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
runners: dfu-util: Convert to -i/--dev-id

Remove the previous dfu-util-specific --pid option and switch to the
common -i/--dev-id one.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
runners: pyocd: Convert to -i/--dev-id

Remove the previous pyocd-specific --board-id option and switch to
the common -i/--dev-id one.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

henrikbrixandersen reviewed (changes requested):
Not sure if this is the right approach for all runners. Take for instance the CANopen Node ID; Everybody working with CANopen knows what a Node ID is, but would likely not associate `--dev-id` with being that from the generic description given. Compare it to e.g. an IP address of a remote debugger - it would not be obvious to pass that in as `--dev-id`? Same applies to serial numbers, USB VID:PID, board names, etc.

Blocking this to ensure we have a discussion about this.

carlescufi commented:
> Not sure if this is the right approach for all runners. Take for instance the CANopen Node ID; Everybody working with CANopen knows what a Node ID is, but would likely not associate `--dev-id` with being that from the generic description given. Compare it to e.g. an IP address of a remote debugger - it would not be obvious to pass that in as `--dev-id`? Same applies to serial numbers, USB VID:PID, board names, etc.

It's a matter of taste more than anything else. Personally I prefer having a single command-line option that is reused by all runners albeit with slight different meanings based on the context, but the opposite argument can be made. For me, anything that identifies a device can fall under the `-i/--dev-id` usage, and I am more comfortable using `-i <node id/vid:pid/serial>` than having to remember specific options for each runner.

mbolivar-nordic commented on scripts/west_commands/runners/core.py:
```diff
@@ -473,7 +473,7 @@ def create(cls, cfg: RunnerConfig,
         caps = cls.capabilities()
         if args.dev_id and not caps.dev_id:
-            _missing_cap(cls, '--dev-id')
+            _missing_cap(cls, '-i/--dev-id')
```
Rebase cruft? Please remove this change.

mbolivar-nordic replied:
> Please remove this change.

Sorry for being vague. I meant "please make this '--dev-id', not '-i/--dev-id, for consistency with the rest of the `_missing_cap` calls". The previous version passed '--dev-id' to `_missing_cap` in one commit, then changed the string to '-i/--dev-id' in a subsequent commit. The first commit was correct.

carlescufi replied:
Oh, I see. I added it because the other ones did not have a short version of the command-line option and wanted the message for the user to be explicit about which option(s) were not supported. But not a big deal, just fixed it now.

mbolivar-nordic commented on scripts/west_commands/runners/nrfjprog.py:
```diff
@@ -73,9 +73,6 @@ def do_add_parser(cls, parser):
         parser.add_argument('--softreset', required=False,
                             action='store_true',
                             help='use reset instead of pinreset')
-        parser.add_argument('--snr', required=False,
-                            help="""Serial number of board to use.
-                            '*' matches one or more characters/digits.""")
```
Let's keep this for compatibility, but feel free to adjust the help text accordingly.

mbolivar-nordic commented on scripts/west_commands/runners/dfu.py:
```diff
@@ -18,12 +18,16 @@
                  dfuse_config=None):
+
+        if not dev_id:
+            raise RuntimeError('Please specify a USB VID:PID with the '
+                               '-i/--dev-id command-line switch.')
```
The error handling in the calling layers expects the constructor to succeed. Please defer raising exceptions to do_run().

mbolivar-nordic commented on scripts/west_commands/runners/dfu.py:
```diff
@@ -42,13 +46,11 @@ def name(cls):
     @classmethod
     def do_add_parser(cls, parser):
         # Required:
-        parser.add_argument("--pid", required=True,
-                            help="USB VID:PID of the board")
```
For compatibility, please change to: `parser.add_argument('--pid', dest='dev_id')`, removing `required=True` and changing the destination attribute in the final args object to match --dev-id's. (you can change `help` to mention the consolidation if you think that'd be good here as well.)

mbolivar-nordic commented on scripts/west_commands/runners/pyocd.py:
```diff
@@ -105,8 +105,6 @@ def do_add_parser(cls, parser):
                                 DEFAULT_PYOCD_TELNET_PORT))
         parser.add_argument('--tui', default=False, action='store_true',
                             help='if given, GDB uses -tui')
-        parser.add_argument('--board-id',
-                            help='ID of board to flash, default is to prompt')
```
Same story here; please preserve compatibility.

mbolivar-nordic reviewed (changes requested):
Thanks for finally crossing off this longstanding TODO.

I defer to @henrikbrixandersen on the canopennode runner.

As for the rest, no objections to adding the option, but there's no need to break compatibility. See comments for details.

mbolivar-nordic commented:
The test cases need updates too.

carlescufi committed:
runners: nrfjprog: Convert to -i/--dev-id

Remove the previous nrfjprog-specific --id option and switch to the
common -i/--dev-id one.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

henrikbrixandersen commented:
I am sorry. I still think these changes are making the command line options _less_ obvious for an end-user. There is no way - except for reading the source code - for en end-user to know what `--dev-id` expects for a given runner. Serial number? Node ID? IP address? IP:port? USB VID:PID?

Unifying these into one also means that adding more ways of selecting a given instance is even more obscure. Say, a runner that supports selection via either serial number or USB VID:PID.

mbolivar-nordic commented:
> There is no way - except for reading the source code - for en end-user to know what `--dev-id` expects for a given runner. Serial number? Node ID? IP address? IP:port? USB VID:PID?

We can fix this in both the .rst files and in the command line help.

> Unifying these into one also means that adding more ways of selecting a given instance is even more obscure. Say, a runner that supports selection via either serial number or USB VID:PID.

Do you actually have a such a situation? I can imagine IP vs. USB debugging with a J-Link, but that seems solvable with either some additional smarts in the runner ("is this argument an IP:port? Then it's IP.  Is it just a number? Then it's USB.") or another option like `--id-format`.

Assuming we sort these things out, are you still opposed to the idea completely, or are you saying you have reservations about the current status?

carlescufi commented:
>We can fix this in both the .rst files and in the command line help.

I completely agree with @mbolivar-nordic. Documenting this properly and then getting the users to understand that a command-line option might take different formats based on the runner seems simpler for the user to me than having to remember a bunch of different ones. But again, this is purely cosmetic and so subject to individual taste.

@henrikbrixandersen an alternative to generalizing it entirely would be to do this only for those that identify a particular debugger instance. So only jlink, nrfjprog and pyocd would use `-i/--dev-id`.

mbolivar-nordic commented:
I think it's time to document the individual runner backends on the .rst side. On the help string side, I think core.py should set the help text for the dev_id option using a self.dev_id_help attribute which subclasses can override. Then each runner can explain what the individual ID contents are via an override of that attribute.

galak commented:
Would using `-u` like pyocd be any clearier?

```
  -u UNIQUE_ID, --uid UNIQUE_ID, --probe UNIQUE_ID
                        Choose a probe by its unique ID or a substring thereof. Optionally prefixed with
                        '<probe-type>:' where <probe-type> is the name of a probe plugin.
```

henrikbrixandersen commented:
> Would using `-u` like pyocd be any clearier?

Not as I see it. What I questioning here is what the benefit by combining these different configuration options into one and leave the user to parse a help text/source code in order to find out what kind of argument the given runner expects when specifying that option.

I am all for combining options for common parameters (e.g. `--serial-number` or `--remote-host` or similiar`), but I fail to see the benefit of combining them into one, when they do not accept the same arguments. 

If I do a `west -r canopen -H` and see e.g. `--node-id` I don't need to read any more help text in order to figure out what to pass as argument. If I run the same command and see `--dev-id` or `--uid` I would not know what to pass without reading the help text.

We could even (I don't believe we have that today?) have bash completion of runner specific arguments, which would make it even more intuitive to use. Type in `west -r canopen --<TAB>`, see `--node-id` in the list and from the option name already there know what to pass as argument - without even looking at the help output or other documentation.

Just to clarify: The above goes for other runner device selection arguments than the CANopen runner `--node-id`, which is just used as an example.

galak commented:
dev-review (Aug 12, 2021):
* suggestion is to support both the old and new option.  All runners for debug/flash should support the new option at min going forward.  Docs/usage help to clarify.

carlescufi commented:
>I think core.py should set the help text for the dev_id option using a self.dev_id_help attribute which subclasses can override

This sounds like a very good idea. I will start with that, given that anyway we are keeping the old options I can actually reuse the same text for both on each runner.

carlescufi commented:
@mbolivar-nordic 

>I think it's time to document the individual runner backends on the .rst side

I am not sure about this one for this PR, but I am happy to discuss this with you to see if we can make it happen.

>I think core.py should set the help text for the dev_id option using a self.dev_id_help attribute which subclasses can override. Then each runner can explain what the individual ID contents are via an override of that attribute.

This is now done.

@henrikbrixandersen 

>suggestion is to support both the old and new option.

I did that for the runners that are controversial.

gmarull commented on scripts/west_commands/runners/nrfjprog.py:
```diff
@@ -73,9 +79,8 @@ def do_add_parser(cls, parser):
-        parser.add_argument('--snr', required=False,
-                            help="""Serial number of board to use.
-                            '*' matches one or more characters/digits.""")
+        parser.add_argument('--snr', required=False, dest='dev_id',
+                            help='Deprecated: use -i/--dev-id instead')
```
What about adding a deprecation warning? This will make deprecation clearly visible allowing people to migrate to `--dev-id`. It can be achieved with something like:

```python
import argparse
import warnings


class DeprecateAction(argparse.Action):
    def __call__(self, parser, namespace, values, option_string=None):
        warnings.warn("Argument %s is deprecated" % self.option_strings)
        setattr(namespace, self.dest, values)


parser = argparse.ArgumentParser()
parser.add_argument("--dev-id", dest="dev_id")
parser.add_argument("--snr", dest="dev_id", action=DeprecateAction,
                    help="[DEPRECATED] Use --dev-id instead")

args = parser.parse_args()
print(args)
```

carlescufi replied:
Thanks @gmarull. I agree this is a welcome addition. Since I am unsure about @mbolivar-nordic's stance on this will be, I will add it as separate commit on top.

gmarull reviewed (approved):
LGTM, I've added one comment. I think deprecation notice would be a useful addition.

henrikbrixandersen commented on scripts/west_commands/runners/canopen_program.py:
```diff
@@ -63,14 +64,18 @@ def name(cls):
-        # Required:
-        parser.add_argument('--node-id', required=True, help='Node ID')
+    def dev_id_help(cls) -> str:
+        return 'CAN Node ID.'
```
```suggestion
        return 'CANopen Node ID.'
```

henrikbrixandersen replied:
This still says "CAN Node ID". CAN does not have node IDs; CANopen does.

carlescufi replied:
Oops, fixed now!

henrikbrixandersen commented on scripts/west_commands/runners/canopen_program.py:
```diff
@@ -88,14 +93,17 @@ def do_add_parser(cls, parser):
 
     def do_run(self, command, **kwargs):
+        if not self.dev_id:
+            raise RuntimeError('Please specify a node ID with the -i/--dev-id '
```
```suggestion
            raise RuntimeError('Please specify a CANopen node ID with the -i/--dev-id '
```

henrikbrixandersen commented on scripts/west_commands/runners/nrfjprog.py:
```diff
@@ -62,7 +62,13 @@ def name(cls):
+
+    @classmethod
+    def dev_id_help(cls) -> str:
+        return '''Device identifier. Use it to select the J-Link Serial Number
```
What this supposed to say "J-Link"? Or is it the serial number of the nRF board?

carlescufi replied:
All of our boards use a J-Link debugger, and the serial number is associated with the J-Link OB chip, so I think it's a good description.

henrikbrixandersen reviewed (changes requested):
A few minor nits, otherwise looks good, thank you for updating.

carlescufi committed:
west: runners: Add a deprecation action to the core code

In order to allow for further options to be deprecated with minimal
impact, add a deprecation argparse Action and a callable instantiator
that can be used to deprecate options in favor of new ones.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi commented:
@mbolivar-nordic and @henrikbrixandersen could you please take another look?
@gmarull thanks for the suggestion, I implemented the deprecation helper Action

mbolivar-nordic reviewed (changes requested):
Looking good now; just a couple of things. Thanks again.
